api/utils.go

```diff
@@ -103,23 +103,48 @@ func proxyAsync(engine *cluster.Engine, w http.ResponseWriter, r *http.Request,
 	r.URL.Host = engine.Addr
 
 	log.WithFields(log.Fields{"method": r.Method, "url": r.URL}).Debug("Proxy request")
-	resp, err := client.Do(r)
-	if err != nil {
-		return err
-	}
 
-	copyHeader(w.Header(), resp.Header)
-	w.WriteHeader(resp.StatusCode)
-	io.Copy(NewWriteFlusher(w), resp.Body)
+	requestDone := make(chan struct{})
+	go func() {
+		defer close(requestDone)
+		var resp *http.Response
+		resp, err = client.Do(r)
+		if err != nil {
+			return
+		}
+		// cleanup
+		defer resp.Body.Close()
 
-	if callback != nil {
-		callback(resp)
-	}
+		copyHeader(w.Header(), resp.Header)
+		w.WriteHeader(resp.StatusCode)
+		io.Copy(NewWriteFlusher(w), resp.Body)
 
-	// cleanup
-	resp.Body.Close()
+		if callback != nil {
+			callback(resp)
+		}
+	}()
 
-	return nil
+	type requestCanceler interface {
+		CancelRequest(*http.Request)
+	}
+
+	var closeNotify <-chan bool
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
+		closeNotify = closeNotifier.CloseNotify()
+	}
+	select {
+	case <-closeNotify:
+		log.Debugf("user connection closed")
+		if rc, ok := client.Transport.(requestCanceler); ok {
+			log.Debugf("cancel request")
+			rc.CancelRequest(r)
+		}
+		// wait for request finish
+		<-requestDone
+		return err
```

Comment:
This one is a little tricky. The common use cases are long running streams like `docker events` or `docker logs`. There is no error when requests finish or user interrupt. So I'm returning `err` from `resp, err = client.Do(r)`.